hir/src/value.rs

```diff
@@ -38,18 +38,28 @@ pub enum ValueData {
         block: Block,
         span: SourceSpan,
     },
+    /// Value is an alias of another value.
+    /// An alias value can't be linked as an instruction result or block parameter. It is used as a
+    /// placeholder when the original instruction or block has been rewritten or modified.
+    Alias {
```

Comment:
> Could you elaborate on the semantics of this value type? I gather from comment that it basically acts like a reference to another `Value`, but this potentially impacts a number of places where we have to make sure that we preserve the semantics of the IR, so I want to make sure I fully understand it, and that we update any affected places in the code.
> 
> Namely, I have the following questions:
> 
> 1. When do we expect to observe `Alias` values?

When we search for a var definition in predecessors, we create a block parameter as a sentinel (in `SSABuilder::find_var`). When all the predecessors use a single value to represent our variable, we create an alias to that value and remove the block parameter (in `SSABuilder::finish_predecessor_lookup`) otherwise, we keep the block parameter. This is the only place where we create an alias. 

> 2. What should be done with an `Alias` value if observed during analysis or transformation passes?

It should be resolved to the underlying value (via `DataFlowGraph::resolve_aliases`).

> 
> Regarding my second question, I'm particularly thinking of the block inlining and treeify passes which clone/modify blocks and instructions, and accordingly, the values they define/use. I just fixed a bug I accidentally introduced by shallow cloning instructions which resulted in cloned instructions placed in different blocks accidentally sharing their ValueLists. Based on the description here, I would assume it is a bug to observe an `Alias` value in most situations, but it isn't clear to me when it should be used or not used; observed or not observed.

It's not a bug, but I can see how it might complicate things down the pipeline. I can get rid of the `Alias` and keep the block parameter instead.

Reply:
Keeping the block parameter is fine, as it will be removed by later transformation passes if it's not needed. But you should also just be able to reference the original `Value`, i.e. the `Alias` value type doesn't sound like it's actually needed, since a `Value` isn't tied to a specific block. As long as the use of a `Value` is dominated by it's definition, you can just use the original `Value` directly (i.e. all control flow paths from the entry block to the block containing the use, flow through the definition, whether that's a block parameter or an instruction result). To illustrate, consider the following IR:

```
pub fn foo(u32, u32) -> u32 {
entry(v0, v1):
    v2 = const.u32 0
    v3 = eq v1, 0
    condbr v3, blk0(v2), blk1

blk0(v4: u32):
    ret v4

blk1:
    v5 = const.u32 2
    v6 = eq v0, 0
    condbr v6, blk2, blk3

blk2:
    br blk3

blk3:
    v7 = div v0, v1
    v8 = mul v7, v5
    br blk4(v8)

blk4(v9: u32):
    ret v9
```

In this function, the use of `v0`, `v1`, and `v5` in `blk3` are safe, because all paths to `blk3` pass through the definition of those values (`v0` and `v1` trivially dominate every block in the function, since they are parameters of the entry block, and `v5` is defined in `blk1`, the immediate dominator of `blk3`.

Note that the block parameter in `blk4` is not needed, but is also perfectly valid, it just obfuscates the origin of `v9`, but it can also be useful in some cases to introduce such parameters, in order to break up the live range of a value.

So coming back to the purpose of `Alias`, it seems like you would intend to use it in `blk3` for the uses of `v0`, `v1`, and `v5`, if I understand correctly. Hopefully my example IR demonstrates why that wouldn't be necessary though. That said, if I'm missing a key difference here, perhaps an example like mine above would help illustrate the use case a bit better for me.

Reply:
Thank you for the code example and detailed explanation! You are right. I removed `ValueData::Alias`.